File: iteration2/src/AdvisorInterface.java

import java.util.ArrayList;
import java.util.Scanner;

public class AdvisorInterface implements Schedule {

    private Scanner scanner;
    private Session session;
    private NotificationsInterface notificationsInt;
    private LoginInterface loginInt;
    private AdvisorCourseRegistrationInterface advCourseRegInt;

    public AdvisorInterface(Session session, LoginInterface loginInt) {
        this.session = session;
        this.loginInt = loginInt;
        advCourseRegInt = new AdvisorCourseRegistrationInterface(session, this);
    }

    // the terminal inteface for advisor
    public void advMenu() {

        while (true) {
            scanner = new Scanner(System.in);
            System.out.println(Colors.RED + "\n--------------------Advisor Menu--------------------\n" + Colors.RESET);
            System.out.println(Colors.YELLOW + "1" + Colors.RESET + ".   View Notifications");
            System.out.println(Colors.YELLOW + "2" + Colors.RESET + ".   View Weekly Schedule");
            System.out.println(Colors.YELLOW + "3" + Colors.RESET + ".   View Given Courses");
            System.out.println(Colors.YELLOW + "4" + Colors.RESET + ".   Course Registration System");
            System.out.println(Colors.YELLOW + "*" + Colors.RESET + ".   Logout");
            System.out.println(Colors.YELLOW + "x" + Colors.RESET + ".   Exit");

            System.out.print("\n" + Colors.BLUE + "--> " + Colors.RESET + "What do you want to do?   ");

            System.out.print(Colors.BLUE);
            char caseToken = scanner.next().charAt(0);
            System.out.print(Colors.RESET);

            switch (caseToken) {
                case '1': // view notifications
                    notificationsInt = new NotificationsInterface(session);
                    notificationsInt.notificationsMenu();
                    break;
                case '2': // view weekly schedule
                    showWeeklySchedule(calculateWeeklySchedule());
                    break;
                case '3': // view given courses
                    showGivenCourses(calculateWeeklySchedule());
                    break;
                case '4': // go to course registration system
                    advCourseRegInt.advRegMenu();
                    break;
                case '*': // logout
                    loginInt.logout();
                    break;
                case 'x': // exit
                    loginInt.exit();
                    break;
                default: // invalid input
                    System.out.println(Colors.YELLOW + "Invalid input! Please try again." + Colors.RESET);
                    advMenu();
                    break;
            }
        }
    }

    // it calculates the weekly schedule of advisor and stores it
    @Override
    public ArrayList<Course> calculateWeeklySchedule() {
        ArrayList<Course> weeklySchedule = ((Lecturer) session.getUser()).getGivenCourses();
        return weeklySchedule;
    }

    // it prints the weekly schedule of advisor
    @Override
    public void showWeeklySchedule(ArrayList<Course> courses) {

        scanner = new Scanner(System.in);

        // creating arraylists for each day of the week
        ArrayList<String> mondayCoursesID = new ArrayList<String>();
        ArrayList<String> tuesdayCoursesID = new ArrayList<String>();
        ArrayList<String> wednesdayCoursesID = new ArrayList<String>();
        ArrayList<String> thursdayCoursesID = new ArrayList<String>();
        ArrayList<String> fridayCoursesID = new ArrayList<String>();
        ArrayList<String> saturdayCoursesID = new ArrayList<String>();
        ArrayList<String> sundayCoursesID = new ArrayList<String>();

        ArrayList<String> mondayCoursesStartTime = new ArrayList<String>();
        ArrayList<String> tuesdayCoursesStartTime = new ArrayList<String>();
        ArrayList<String> wednesdayCoursesStartTime = new ArrayList<String>();
        ArrayList<String> thursdayCoursesStartTime = new ArrayList<String>();
        ArrayList<String> fridayCoursesStartTime = new ArrayList<String>();
        ArrayList<String> saturdayCoursesStartTime = new ArrayList<String>();
        ArrayList<String> sundayCoursesStartTime = new ArrayList<String>();

        ArrayList<String> mondayCoursesPlace = new ArrayList<String>();
        ArrayList<String> tuesdayCoursesPlace = new ArrayList<String>();
        ArrayList<String> wednesdayCoursesPlace = new ArrayList<String>();
        ArrayList<String> thursdayCoursesPlace = new ArrayList<String>();
        ArrayList<String> fridayCoursesPlace = new ArrayList<String>();
        ArrayList<String> saturdayCoursesPlace = new ArrayList<String>();
        ArrayList<String> sundayCoursesPlace = new ArrayList<String>();

        System.out.println(Colors.RED
                + "\n-------------------------------------------------------------Weekly Schedule--------------------------------------------------------------\n"
                + Colors.RESET);
        System.out.println(
                "------------------------------------------------------------------------------------------------------------------------------------------");

        System.out.printf("|  %-15s|  %-14s|  %-14s|  %-14s|  %-14s|  %-14s|  %-14s|  %-14s|%n", "", "Monday",
                "Tuesday", "Wednesday",
                "Thursday", "Friday", "Saturday", "Sunday");

        Course course;
        CourseSession courseSession;

        for (int i = 0; i < courses.size(); i++) {
            course = courses.get(i);
            courseSession = course.getCourseSession();

            for (int j = 0; j < courseSession.getCourseDay().size(); j++) {
                if (courseSession.getCourseDay().get(j).equals("Pazartesi")) {
                    if (course instanceof Lecture) {
                        mondayCoursesID.add(((Lecture) course).getLectureID());
                    } else {
                        mondayCoursesID.add(((Lab) course).getLabID());
                    }
                    mondayCoursesPlace.add(courseSession.getCoursePlace().get(j));
                    mondayCoursesStartTime.add(courseSession.getCourseStartTime().get(j));

                } else if (courseSession.getCourseDay().get(j).equals("Salı")) {
                    if (course instanceof Lecture) {
                        tuesdayCoursesID.add(((Lecture) course).getLectureID());
                    } else {
                        tuesdayCoursesID.add(((Lab) course).getLabID());
                    }

                    tuesdayCoursesPlace.add(courseSession.getCoursePlace().get(j));
                    tuesdayCoursesStartTime.add(courseSession.getCourseStartTime().get(j));

                } else if (courseSession.getCourseDay().get(j).equals("Çarşamba")) {
                    if (course instanceof Lecture) {
                        wednesdayCoursesID.add(((Lecture) course).getLectureID());
                    } else {
                        wednesdayCoursesID.add(((Lab) course).getLabID());
                    }
                    wednesdayCoursesPlace.add(courseSession.getCoursePlace().get(j));
                    wednesdayCoursesStartTime.add(courseSession.getCourseStartTime().get(j));

                } else if (courseSession.getCourseDay().get(j).equals("Perşembe")) {
                    if (course instanceof Lecture) {
                        thursdayCoursesID.add(((Lecture) course).getLectureID());
                    } else {
                        thursdayCoursesID.add(((Lab) course).getLabID());
                    }
                    thursdayCoursesPlace.add(courseSession.getCoursePlace().get(j));
                    thursdayCoursesStartTime.add(courseSession.getCourseStartTime().get(j));

                } else if (courseSession.getCourseDay().get(j).equals("Cuma")) {
                    if (course instanceof Lecture) {
                        fridayCoursesID.add(((Lecture) course).getLectureID());
                    } else {
                        fridayCoursesID.add(((Lab) course).getLabID());
                    }

                    fridayCoursesPlace.add(courseSession.getCoursePlace().get(j));
                    fridayCoursesStartTime.add(courseSession.getCourseStartTime().get(j));

                } else if (courseSession.getCourseDay().get(j).equals("Cumartesi")) {
                    if (course instanceof Lecture) {
                        saturdayCoursesID.add(((Lecture) course).getLectureID());
                    } else {
                        saturdayCoursesID.add(((Lab) course).getLabID());
                    }
                    saturdayCoursesPlace.add(courseSession.getCoursePlace().get(j));
                    saturdayCoursesStartTime.add(courseSession.getCourseStartTime().get(j));

                } else if (courseSession.getCourseDay().get(j).equals("Pazar")) {
                    if (course instanceof Lecture) {
                        sundayCoursesID.add(((Lecture) course).getLectureID());
                    } else {
                        sundayCoursesID.add(((Lab) course).getLabID());
                    }
                    sundayCoursesPlace.add(courseSession.getCoursePlace().get(j));
                    sundayCoursesStartTime.add(courseSession.getCourseStartTime().get(j));

                }
            }

        }
        String mondayCourses;
        String tuesdayCourses;
        String wednesdayCourses;
        String thursdayCourses;
        String fridayCourses;
        String saturdayCourses;
        String sundayCourses;

        String mondayCoursePlace;
        String tuesdayCoursePlace;
        String wednesdayCoursePlace;
        String thursdayCoursePlace;
        String fridayCoursePlace;
        String saturdayCoursePlace;
        String sundayCoursePlace;

        WeeklySchedule weeklySchedule = new WeeklySchedule();

        for (int k = 0; k < CourseSessionTimes.SESSION_START.size(); k++) {

            mondayCourses = weeklySchedule.printMondayCourses(mondayCoursesID, mondayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            tuesdayCourses = weeklySchedule.printTuesdayCourses(tuesdayCoursesID, tuesdayCoursesStartTime,
                    CourseSessionTimes.SESSION_START,
                    k);
            wednesdayCourses = weeklySchedule.printWednesdayCourses(wednesdayCoursesID, wednesdayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            thursdayCourses = weeklySchedule.printThursdayCourses(thursdayCoursesID, thursdayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            fridayCourses = weeklySchedule.printFridayCourses(fridayCoursesID, fridayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            saturdayCourses = weeklySchedule.printSaturdayCourses(saturdayCoursesID, saturdayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            sundayCourses = weeklySchedule.printSundayCourses(sundayCoursesID, sundayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);

            mondayCoursePlace = weeklySchedule.printMondayCoursePlace(mondayCoursesPlace, mondayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            tuesdayCoursePlace = weeklySchedule.printTuesdayCoursePlace(tuesdayCoursesPlace, tuesdayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            wednesdayCoursePlace = weeklySchedule.printWednesdayCoursePlace(wednesdayCoursesPlace,
                    wednesdayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            thursdayCoursePlace = weeklySchedule.printThursdayCoursePlace(thursdayCoursesPlace,
                    thursdayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            fridayCoursePlace = weeklySchedule.printFridayCoursePlace(fridayCoursesPlace, fridayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);
            saturdayCoursePlace = weeklySchedule.printSaturdayCoursePlace(saturdayCoursesPlace,
                    saturdayCoursesStartTime,
                    CourseSessionTimes.SESSION_START,
                    k);
            sundayCoursePlace = weeklySchedule.printSundayCoursePlace(sundayCoursesPlace, sundayCoursesStartTime,
                    CourseSessionTimes.SESSION_START, k);

            if (mondayCourses != "" || tuesdayCourses != "" || wednesdayCourses != "" || thursdayCourses != ""
                    || fridayCourses != "" || saturdayCourses != "" || sundayCourses != "") {
                System.out.println(
                        "------------------------------------------------------------------------------------------------------------------------------------------");
                System.out.printf("|  %-15s|  %-14s|  %-14s|  %-14s|  %-14s|  %-14s|  %-14s|  %-14s|%n",
                        "", mondayCourses,
                        tuesdayCourses, wednesdayCourses, thursdayCourses, fridayCourses, saturdayCourses,
                        sundayCourses);
                System.out.printf("|  %-15s|  %-14s|  %-14s|  %-14s|  %-14s|  %-14s|  %-14s|  %-14s|%n",
                        CourseSessionTimes.SESSION_START.get(k) + " - " + CourseSessionTimes.SESSION_END.get(k),
                        mondayCoursePlace, tuesdayCoursePlace, wednesdayCoursePlace, thursdayCoursePlace,
                        fridayCoursePlace, saturdayCoursePlace, sundayCoursePlace);
            }

        }
        System.out.println(
                "------------------------------------------------------------------------------------------------------------------------------------------\n");

        System.out.println(Colors.YELLOW + "0" + Colors.RESET + ".  Back to Advisor Menu");
        System.out.print("\n" + Colors.BLUE + "--> " + Colors.RESET + "What do you want to do?   ");
        System.out.print(Colors.BLUE);
        char caseToken = scanner.next().charAt(0);
        System.out.print(Colors.RESET);
        switch (caseToken) {
            case '0':
                advMenu();
                break;
            default:
                System.out.println(Colors.YELLOW + "Invalid input! Please try again." + Colors.RESET);
                showWeeklySchedule(courses);
                break;
        }

    }

    // it prints the given courses of advisor
    public void showGivenCourses(ArrayList<Course> courses) {
        scanner = new Scanner(System.in);
        System.out.println(Colors.RED + "\n----------------------Given Courses----------------------\n" + Colors.RESET);

        Course course;
        ArrayList<String> courseID = new ArrayList<String>();

        System.out.println("---------------------------------------------------------");
        System.out.printf("|  %s%-13s%s|  %s%-35s%s  |%n", Colors.YELLOW, "Course ID", Colors.RESET, Colors.YELLOW,
                "Course Name", Colors.RESET);

        for (int i = 0; i < courses.size(); i++) {
            course = courses.get(i);
            if (course instanceof Lecture) {
                courseID.add(((Lecture) course).getLectureID());
            } else {
                courseID.add(((Lab) course).getLabID());
            }

            System.out.println("---------------------------------------------------------");
            System.out.printf("|  %-13s|  %-35s  |%n", courseID.get(i), course.getCourseName());

        }
        System.out.println("---------------------------------------------------------\n");

        System.out.println(Colors.YELLOW + "0" + Colors.RESET + ".  Back to Advisor Menu");
        System.out.print("\n" + Colors.BLUE + "--> " + Colors.RESET + "What do you want to do?   ");
        System.out.print(Colors.BLUE);
        char caseToken = scanner.next().charAt(0);
        System.out.print(Colors.RESET);
        switch (caseToken) {
            case '0':
                advMenu();
                break;
            default:
                System.out.println(Colors.YELLOW + "Invalid input! Please try again." + Colors.RESET);
                showGivenCourses(courses);
                break;
        }

    }

}

File: iteration3/src/advisor_interface.py
from advisor_course_registration_interface import AdvisorCourseRegistrationInterface
from notifications_interface import NotificationsInterface
from session import Session
from lecture import Lecture


class AdvisorInterface:
    def __init__(self, session: Session):
        self.__session = session
        self.__adv_course_reg_int = AdvisorCourseRegistrationInterface(session=self.__session)

    def adv_menu(self):
        while True:
            print("\n> Advisor Menu")
            print("1. View Notifications")
            print("2. View Weekly Schedule")
            print("3. View Given Courses")
            print("4. Course Registration System")
            print("*. Logout")
            print("x. Exit")
            choice = input("\nWhat do you want to do? ")

            if choice == "1":
                notifications_int = NotificationsInterface(self.__session)
                notifications_int.notifications_menu()
            elif choice == "2":
                weekly_schedule = self.__calculate_weekly_schedule()
                self.__show_weekly_schedule(weekly_schedule)
            elif choice == "3":
                given_courses = self.__calculate_weekly_schedule()
                self.__show_given_courses(given_courses)
            elif choice == "4":
                self.__adv_course_reg_int.adv_reg_menu()
            elif choice == "*":
                self.__login_int.logout()
            elif choice == "x":
                self.__login_int.exit()
            else:
                print("Invalid input! Please try again.")

    def __calculate_weekly_schedule(self):
        # Assuming `session` contains the current session and `get_user()` returns the current user
        weekly_schedule = self.__session.get_user().get_given_courses()
        return weekly_schedule

    def __show_weekly_schedule(self, courses):
        # initializing lists for each day of the week
        (
            monday_courses_id,
            tuesday_courses_id,
            wednesday_courses_id,
            thursday_courses_id,
            friday_courses_id,
            saturday_courses_id,
            sunday_courses_id,
        ) = ([] for _ in range(7))
        (
            monday_courses_start_time,
            tuesday_courses_start_time,
            wednesday_courses_start_time,
            thursday_courses_start_time,
            friday_courses_start_time,
            saturday_courses_start_time,
            sunday_courses_start_time,
        ) = ([] for _ in range(7))
        (
            monday_courses_place,
            tuesday_courses_place,
            wednesday_courses_place,
            thursday_courses_place,
            friday_courses_place,
            saturday_courses_place,
            sunday_courses_place,
        ) = ([] for _ in range(7))

        print("\n>> Weekly Schedule")
        print(
            "------------------------------------------------------------------------------------------------------------------------------------------"
        )
        print(
            "|  {:<15}|  {:<14}|  {:<14}|  {:<14}|  {:<14}|  {:<14}|  {:<14}|  {:<14}|".format(
                "",
                "Monday",
                "Tuesday",
                "Wednesday",
                "Thursday",
                "Friday",
                "Saturday",
                "Sunday",
            )
        )

        for course in courses:
            course_session = course.get_course_session()
            for j, day in enumerate(course_session.get_course_day()):
                if day == "Pazartesi":
                    if isinstance(course, Lecture):
                        monday_courses_id.append(course.get_lecture_id())
                    else:
                        monday_courses_id.append(course.get_lab_id())
                    monday_courses_place.append(course_session.get_course_place()[j])
                    monday_courses_start_time.append(
                        course_session.get_course_start_time()[j]
                    )

                elif day == "Salı":
                    if isinstance(course, Lecture):
                        tuesday_courses_id.append(course.get_lecture_id())
                    else:
                        tuesday_courses_id.append(course.get_lab_id())
                    tuesday_courses_place.append(course_session.get_course_place()[j])
                    tuesday_courses_start_time.append(
                        course_session.get_course_start_time()[j]
                    )

                elif day == "Çarşamba":
                    if isinstance(course, Lecture):
                        wednesday_courses_id.append(course.get_lecture_id())
                    else:
                        wednesday_courses_id.append(course.get_lab_id())
                    wednesday_courses_place.append(course_session.get_course_place()[j])
                    wednesday_courses_start_time.append(
                        course_session.get_course_start_time()[j]
                    )

        while True:
            print("0.  Back to Advisor Menu")
            choice = input("\nWhat do you want to do? ")
            if choice == "0":
                return
            else:
                print("Invalid input! Please try again.")

    def __show_given_courses(self, courses):
        course_ids = []
        print("\n>> Given Courses")
        print("---------------------------------------------------------")
        print("|  Course ID  |  Course Name")

        for course in courses:
            if isinstance(course, Lecture):
                course_ids.append(course.get_lecture_id())
            else:
                course_ids.append(course.get_lab_id())

            print("---------------------------------------------------------")
            print(f"|  {course_ids[-1]:<11}|  {course.get_course_name():<35}")

        print("---------------------------------------------------------\n")

        while True:
            print("0.  Back to Advisor Menu")
            choice = input("\nWhat do you want to do? ")

            if choice == "0":
                return
            else:
                print("Invalid input! Please try again.")
